File: survey/drivers/package_url/packageurl_js/parse.mjs
#!/usr/bin/env node

import { PackageURL } from "packageurl-js";
import { createInterface } from "node:readline";

const rl = createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false,
});

for await (let line of rl) {
    line = line.trim();
    if (line === "") {
        continue;
    }
    try {
        const purl = PackageURL.fromString(line);
        console.log(JSON.stringify({
            "type": purl.type,
            "name": purl.name,
            "namespace": purl.namespace,
            "version": purl.version,
            "qualifiers": purl.qualifiers,
            "subpath": purl.subpath,
        }))
    } catch (e) {
        console.log(JSON.stringify({"error": e.toString()}));
    }
}

File: survey/drivers/package_url/packageurl_ruby/parse.rb
#!/usr/bin/env ruby

require 'json'

require_relative 'repo/lib/package_url'

STDIN.each_line {|line|
    line.chomp!
    return if line.empty?

    begin
        purl = PackageURL.parse(line)

        puts JSON.generate({
            type: purl.type,
            name: purl.name,
            namespace: purl.namespace,
            version: purl.version,
            qualifiers: purl.qualifiers,
            subpath: purl.subpath,
        })
    rescue => error
        puts JSON.generate({
            error: error.to_s,
        })
    end
    STDOUT.flush
}
